go/common/entity/entity.go

```diff
@@ -53,6 +53,59 @@ type Entity struct { // nolint: maligned
 	Nodes []signature.PublicKey `json:"nodes,omitempty"`
 }
 
+// UnmarshalCBOR is a custom deserializer that handles both v1 and v2 Entity
+// structures.  A v1 structure is converted to v2 seamlessly if the field
+// AllowEntitySignedNodes is false or missing, otherwise an error is returned.
+func (e *Entity) UnmarshalCBOR(data []byte) error {
+	// Determine Entity structure version.
+	v, err := cbor.GetVersion(data)
+	if err != nil {
+		return err
+	}
+	switch v {
+	case 1:
+		// Old version had an extra field that was used only for debugging/tests.
+		type EntityV1 struct { // nolint: maligned
+			cbor.Versioned
+			ID                     signature.PublicKey   `json:"id"`
+			Nodes                  []signature.PublicKey `json:"nodes,omitempty"`
+			AllowEntitySignedNodes bool                  `json:"allow_entity_signed_nodes,omitempty"`
+		}
+		var ev1 EntityV1
+		if err = cbor.Unmarshal(data, &ev1); err != nil {
+			return err
+		}
+		// Make sure that AllowEntitySignedNodes is not enabled.
+		if ev1.AllowEntitySignedNodes {
+			return fmt.Errorf("entity descriptor must have allow_entity_signed_nodes set to false")
+		}
+		// Convert into new format.
+		e.Versioned = cbor.NewVersioned(2)
+		e.ID = ev1.ID
+		e.Nodes = ev1.Nodes
+		return nil
+	case 2:
+		// New version, call the default unmarshaler.
+		// Unfortunately, we can't just use the existing Entity struct, because
```

Comment:
I _think_ you could also do
```
case 2:
    type e2 Entity
    return cbor.Unmarshal(data, (*e2)(e))
```

Reply:
No, that doesn't work, see coment above -- it causes an infinite loop :)

Reply:
Hm, i think it should, unless i'm doing something weird:

```
root@f23e01d9066c:/code/go/common/entity# git diff origin/andrej/bugfix/entity-v2-again
diff --git a/go/common/entity/entity.go b/go/common/entity/entity.go
index c9d61a5c4..76aa5d430 100644
--- a/go/common/entity/entity.go
+++ b/go/common/entity/entity.go
@@ -85,22 +85,8 @@ func (e *Entity) UnmarshalCBOR(data []byte) error {
                e.Nodes = ev1.Nodes
                return nil
        case 2:
-               // New version, call the default unmarshaler.
-               // Unfortunately, we can't just use the existing Entity struct, because
-               // it would cause an infinite loop when calling Unmarshal.
-               type EntityV2 struct { // nolint: maligned
-                       cbor.Versioned
-                       ID    signature.PublicKey   `json:"id"`
-                       Nodes []signature.PublicKey `json:"nodes,omitempty"`
-               }
-               var ev2 EntityV2
-               if err = cbor.Unmarshal(data, &ev2); err != nil {
-                       return err
-               }
-               e.Versioned = ev2.Versioned
-               e.ID = ev2.ID
-               e.Nodes = ev2.Nodes
-               return nil
+               type e2 Entity
+               return cbor.Unmarshal(data, (*e2)(e))
        default:
                return fmt.Errorf("invalid entity descriptor version: %v", v)
        }


root@f23e01d9066c:/code/go/common/entity# go test
PASS
ok  	github.com/oasisprotocol/oasis-core/go/common/entity	0.004s

```

Reply:
OK, that's weird, it passes for me now as well :man_shrugging:  I'll update the code, thanks :)

Reply:
thanks!